pineappl/src/grid.rs

```diff
@@ -1925,6 +2038,50 @@ impl Grid {
             })
             .collect();
     }
+
+    /// Returns `true` if the first initial state needs a convolution, `false` otherwise.
+    pub fn has_pdf1(&self) -> bool {
```

Comment:
I guess you already have it on your list, but just in case: this should be a one-line function just calling another private function `fn has_pdf(&self, &i32) -> bool` as should `has_pdf2` be (i.e. they are both just syntactical sugar) (same for `initial_state_1` and its sister of course)

Reply:
In this case the we'd have more lines after code deduplication than before so I don't consider this necessary (yet). Note that `has_pdf{1,2}` and `initial_state_{1,2}` is also subject to change in `v0.6.0` to support convolutions with fragmentation functions.